src/vuln_analysis/utils/concurrency.py

```
@@ -86,3 +97,91 @@ def get_effective_rate_limiter(local_max_rate: int | None, builder: "Builder") -
+    """
+    lock_file = get_lock_path(resource_path, lock_dir)
+    lock = FileLock(lock_file, timeout=timeout)
+
+    try:
+        logger.debug("Acquiring lock for resource: %s (lock file: %s)", resource_path, lock_file)
+        with lock.acquire(timeout=timeout):
+            logger.debug("Lock acquired for resource: %s", resource_path)
+            yield lock
+            logger.debug("Releasing lock for resource: %s", resource_path)
+    except Timeout:
+        logger.error("Timeout waiting for lock on resource: %s (waited %d seconds)", resource_path, timeout)
```

Ruff (0.14.14)

[warning] Use `logging.exception` instead of `logging.error`

Replace with `exception`

(TRY400)